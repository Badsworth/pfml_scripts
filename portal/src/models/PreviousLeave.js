/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Previous leave period
 */
import BaseModel from "./BaseModel";

class PreviousLeave extends BaseModel {
  get defaults() {
    return {
      is_for_current_employer: null,
      is_for_same_reason_as_leave_reason: null,
      leave_end_date: null,
      leave_minutes: null,
      leave_reason: null, // PreviousLeaveReason
      leave_start_date: null,
      // this ID field is currently only used by the /applications API in the Claimant portal.
      previous_leave_id: null,
      worked_per_week_minutes: null,
    };
  }
}

/**
 * Enums for the Application's `previous_leaves[].leave_reason` field
 * @enum {string}
 */
export const PreviousLeaveReason = {
  activeDutyFamily:
    "Managing family affairs while a family member is on active duty in the armed forces",
  bonding: "Bonding with my child after birth or placement",
  care: "Caring for a family member with a serious health condition",
  medical: "An illness or injury",
  pregnancy: "Pregnancy",
  serviceMemberFamily:
    "Caring for a family member who serves in the armed forces",
  // Unknown may be displayed to Leave Admins, but isn't
  // an option we display to Claimants
  unknown: "Unknown",
};

export default PreviousLeave;
