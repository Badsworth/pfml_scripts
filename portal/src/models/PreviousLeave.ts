/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Previous leave period
 */
import BaseModel from "./BaseModel";

class PreviousLeave extends BaseModel {
  get defaults() {
    return {
      is_for_current_employer: null,
      leave_end_date: null,
      leave_minutes: null,
      leave_reason: null, // PreviousLeaveReason
      leave_start_date: null,
      // this ID field is currently only used by the /applications API in the Claimant portal.
      previous_leave_id: null,
      type: null, // PreviousLeaveType
      worked_per_week_minutes: null,
    };
  }
}

/**
 * Enums for the Application's `previous_leaves[].leave_reason` field
 * @enum {string}
 */
export const PreviousLeaveReason = {
  activeDutyFamily:
    "Managing family affairs while a family member is on active duty in the armed forces",
  bonding: "Bonding with my child after birth or placement",
  care: "Caring for a family member with a serious health condition",
  medical: "An illness or injury",
  pregnancy: "Pregnancy",
  serviceMemberFamily:
    "Caring for a family member who serves in the armed forces",
  // Unknown may be displayed to Leave Admins, but isn't
  // an option we display to Claimants
  unknown: "Unknown",
};

/**
 * Possible values for Application's "previous_leaves[].type" field.
 * @enum {string}
 */
export const PreviousLeaveType = {
  otherReason: "other_reason",
  sameReason: "same_reason",
};

export default PreviousLeave;
