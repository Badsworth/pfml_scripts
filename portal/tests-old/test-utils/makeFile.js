/**
 * Make a File instance
 * @see https://developer.mozilla.org/en-US/docs/Web/API/File
 * @param {object} attrs File attributes
 * @param {string} attrs.name The file's name
 * @param {string} attrs.type The file's MIME type
 * @returns {File}
 */
export const makeFile = (attrs = {}) => {
  const { name, type } = Object.assign(
    {
      name: "file.pdf",
      type: "application/pdf",
    },
    attrs
  );

  return new File([], name, { type });
};

/**
 * Create an HTML <input> for a test's event.target
 * @param {object} attrs - HTML attributes for the input
 * @returns {HTMLInputElement}
 */
export function createInputElement(attrs) {
  const input = document.createElement("input");

  Object.entries(attrs).forEach(([attributeName, value]) => {
    input.setAttribute(attributeName, value);
  });

  if (attrs.checked === false) input.checked = false;

  return input;
}
