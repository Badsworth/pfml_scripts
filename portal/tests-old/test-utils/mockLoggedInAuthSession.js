import { Auth } from "@aws-amplify/auth";

/**
 * Mock the Cognito authentication session so a test user appears logged in
 */
const mockLoggedInAuthSession = () => {
  const accessTokenJwt =
    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiQnVkIn0.YDRecdsqG_plEwM0H8rK7t2z0R3XRNESJB5ZXk-FRN8";

  jest.spyOn(Auth, "currentSession").mockImplementation(() =>
    Promise.resolve({
      accessToken: { jwtToken: accessTokenJwt },
    })
  );
};

export default mockLoggedInAuthSession;
