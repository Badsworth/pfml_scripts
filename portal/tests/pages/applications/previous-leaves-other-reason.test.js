import { renderWithAppLogic, simulateEvents } from "../../test-utils";
import PreviousLeave from "../../../src/models/PreviousLeave";
import PreviousLeavesOtherReason from "../../../src/pages/applications/previous-leaves-other-reason";

jest.mock("../../../src/hooks/useAppLogic");

const setup = (claimAttrs = {}) => {
  const {
    appLogic,
    claim,
    wrapper,
  } = renderWithAppLogic(PreviousLeavesOtherReason, { claimAttrs });

  const { changeRadioGroup, submitForm } = simulateEvents(wrapper);

  return {
    appLogic,
    changeRadioGroup,
    claim,
    submitForm,
    wrapper,
  };
};

describe("PreviousLeavesOtherReason", () => {
  it("renders the page", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("submits form with has_previous_leaves_other_reason value", async () => {
    const { appLogic, changeRadioGroup, claim, submitForm } = setup();
    const spy = jest.spyOn(appLogic.benefitsApplications, "update");

    await changeRadioGroup("has_previous_leaves_other_reason", "true");
    await submitForm();
    expect(spy).toHaveBeenCalledWith(claim.application_id, {
      has_previous_leaves_other_reason: true,
    });
  });

  it("sets previous_leaves_other_reason to null when has_previous_leaves_other_reason is false and previous_leaves exist", async () => {
    const { appLogic, changeRadioGroup, claim, submitForm } = setup({
      previous_leaves_other_reason: [new PreviousLeave()],
    });
    const spy = jest.spyOn(appLogic.benefitsApplications, "update");

    await changeRadioGroup("has_previous_leaves_other_reason", "false");
    await submitForm();
    expect(spy).toHaveBeenCalledWith(claim.application_id, {
      has_previous_leaves_other_reason: false,
      previous_leaves_other_reason: null,
    });
  });

  it("does not set previous_leaves_other_reason to null when has_previous_leaves_other_reason is false but previous_leaves do not exist", async () => {
    const { appLogic, changeRadioGroup, claim, submitForm } = setup({
      previous_leaves_other_reason: [],
    });
    const spy = jest.spyOn(appLogic.benefitsApplications, "update");

    await changeRadioGroup("has_previous_leaves_other_reason", "false");
    await submitForm();
    expect(spy).toHaveBeenCalledWith(claim.application_id, {
      has_previous_leaves_other_reason: false,
    });
  });
});
