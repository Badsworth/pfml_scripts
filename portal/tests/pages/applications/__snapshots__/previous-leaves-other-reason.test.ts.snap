// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviousLeavesOtherReason renders the page 1`] = `
<div>
  <form
    class="usa-form"
    method="post"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-2 font-heading-sm line-height-sans-3"
      tabindex="-1"
    >
      Other leave, benefits, and income
    </h1>
    <fieldset
      class="usa-fieldset margin-top-3 usa-form-group"
    >
      <legend
        class="usa-label text-bold usa-legend font-heading-lg line-height-sans-3 margin-bottom-1 maxw-tablet"
      >
        <h2
          class="font-heading-lg text-bold"
        >
          Did you take any previous leave between January 1, 2021 and January 1, 2022 and for a different qualifying reason?
        </h2>
      </legend>
      <span
        class="display-block line-height-sans-5 measure-5 usa-intro"
      >
        <h3
          class="font-heading-sm text-bold"
        >
          <svg
            aria-hidden="true"
            class="usa-icon usa-icon--size-3 text-secondary text-middle margin-right-1 margin-top-neg-05"
            focusable="false"
            role="img"
          >
            <use
              xlink:href="/img/sprite.svg#check"
            />
          </svg>
          You must report paid or unpaid leave for these qualifying reasons that aren't the reason you are applying for paid leave now:
        </h3>
        <ul
          class="usa-list margin-left-4"
        >
          <li>
            You had a serious health condition, including illness, injury, or pregnancy.
          </li>
          <li>
            If you were sick, you were out of work for at least 3 days and needed continuing care from your health care provider or needed inpatient care.
          </li>
          <li>
            You bonded with your child after birth or placement.
          </li>
          <li>
            You needed to manage family affairs while a family member is on active duty in the armed forces.
          </li>
          <li>
            You needed to care for a family member who serves in the armed forces.
          </li>
          <li>
            You needed to care for a family member with a serious health condition and your leave began on or after July 1, 2021.
          </li>
        </ul>
        <h3
          class="font-heading-sm text-bold"
        >
          <svg
            aria-hidden="true"
            class="usa-icon usa-icon--size-3 text-error text-middle margin-right-1 margin-top-neg-05"
            focusable="false"
            role="img"
          >
            <use
              xlink:href="/img/sprite.svg#close"
            />
          </svg>
          You don't need to report:
        </h3>
        <ul
          class="usa-list margin-left-4"
        >
          <li>
            Leave that was taken through Massachusetts' PFML program
          </li>
        </ul>
      </span>
      <div
        class="margin-top-3"
      >
        <div
          class="usa-radio measure-5 bg-transparent"
        >
          <input
            class="usa-radio__input"
            id="InputChoice1"
            name="has_previous_leaves_other_reason"
            type="radio"
            value="true"
          />
          <label
            class="usa-radio__label"
            for="InputChoice1"
          >
            Yes
            <span
              class="usa-hint text-base-dark"
            >
              <br />
              I have taken previous leave for one or more qualifying reasons
            </span>
          </label>
        </div>
        <div
          class="usa-radio measure-5 bg-transparent"
        >
          <input
            class="usa-radio__input"
            id="InputChoice2"
            name="has_previous_leaves_other_reason"
            type="radio"
            value="false"
          />
          <label
            class="usa-radio__label"
            for="InputChoice2"
          >
            No
            <span
              class="usa-hint text-base-dark"
            >
              <br />
              I haven’t taken previous leave for a qualifying reason, or my other leave was through Massachusetts’ PFML program
            </span>
          </label>
        </div>
      </div>
    </fieldset>
    <button
      class="usa-button position-relative margin-top-4"
      type="submit"
    >
      Save and continue
    </button>
  </form>
</div>
`;
